Forcing builder

sverhoeven opened the pull request:
Use fluent builder pattern to construct esmvaltool recipes.

Recipes for models with esmvaltool recipes are build using builder and existing diagnostic.

TODO
* [x] DefaultForcing
* [x] GenericDistributedForcing
* [x] GenericLumpedForcing
* [x] PCRGlobWBForcing
* [x] LisfloodForcing
* [x] MarrmotForcing
* [x] WflowForcing
* [x] HypeForcing
* [x] tests
* [x] mypy
* [x] flake8
* [x] documentation
* [x] for generic forcing remove '*' from esmvaltool output file names.
* [x] non ERA5 / ERA-Interim dataset

sverhoeven committed:
Make generic distributed esmvaltool recipe

Peter9192 committed:
New API for forcing generator

sverhoeven committed:
Use fluent builder pattern to construct esmvaltool recipes

sverhoeven committed:
Added builder.add_variables(['pr'])

sverhoeven committed:
Merge remote-tracking branch 'origin/main' into forcing_builder

sverhoeven committed:
Working GenericDistributedForcing.generate

sverhoeven committed:
Move all esmvaltool specific stuff to own package

This make base package much cleaner

sverhoeven committed:
Move builder test

sverhoeven committed:
Fix mypy errors

sverhoeven committed:
Use builder for LisfloodForcing

sverhoeven committed:
Dont use name of Python built-in package as module name

Got
Traceback (most recent call last):
  File "/home/verhoes/git/eWaterCycle/ewatercycle/src/ewatercycle/esmvaltool/diagnostic/copy.py", line 7, in <module>
    from esmvaltool.diag_scripts.shared import (
  File "/home/verhoes/mambaforge/envs/ewatercycle/lib/python3.10/site-packages/esmvaltool/diag_scripts/shared/__init__.py", line 2, in <module>
    from . import io, iris_helpers, names, plot
  File "/home/verhoes/mambaforge/envs/ewatercycle/lib/python3.10/site-packages/esmvaltool/diag_scripts/shared/io.py", line 5, in <module>
    from pprint import pformat
  File "/home/verhoes/mambaforge/envs/ewatercycle/lib/python3.10/pprint.py", line 38, in <module>
    import dataclasses as _dataclasses
  File "/home/verhoes/mambaforge/envs/ewatercycle/lib/python3.10/dataclasses.py", line 3, in <module>
    import copy
  File "/home/verhoes/git/eWaterCycle/ewatercycle/src/ewatercycle/esmvaltool/diagnostic/copy.py", line 7, in <module>
    from esmvaltool.diag_scripts.shared import (
ImportError: cannot import name 'ProvenanceLogger' from partially initialized module 'esmvaltool.diag_scripts.shared' (most likely due to a circular import) (/home/verhoes/mambaforge/envs/ewatercycle/lib/python3.10/site-packages/esmvaltool/diag_scripts/shared/__init__.py)

sverhoeven committed:
Added GenericLumpedForcing, untested as dcache is curently down

sverhoeven committed:
Use recipe builder for MarrmotForcing

sverhoeven committed:
Use builder in WflowForcing

sverhoeven committed:
Use builder for ForcingHype

sverhoeven committed:
Merge remote-tracking branch 'origin/main' into default-gridded-forcing

sverhoeven committed:
Merge branch 'default-gridded-forcing' into forcing_builder

BSchilperoort commented:
I tried to use the following dataset definition to use CMIP data (as that can be automatically downloaded by esmvaltool. It failed on this error:

<details><summary>Error</summary>


```py
File [~/micromamba/envs/ewatercycle310/lib/python3.10/site-packages/esmvalcore/_recipe/check.py:110](https://file+.vscode-resource.vscode-cdn.net/home/bart/git/ewatercycle/~/micromamba/envs/ewatercycle310/lib/python3.10/site-packages/esmvalcore/_recipe/check.py:110), in variable(var, required_keys)
    104 missing = required - set(var)
    105 if missing:
    106     raise RecipeError(
    107         f"Missing keys {missing} in\n"
    108         f"{pformat(var)}\n"
    109         "for variable {var['variable_group']} in diagnostic "
--> 110         f"{var['diagnostic']}")

KeyError: 'diagnostic'
```

</details> 


<details><summary>Dataset</summary>

```py
from ewatercycle.esmvaltool.models import Dataset
cmip_dataset = Dataset(**{
    "dataset": "EC-Earth3",
    "project": "CMIP6", 
    "grid": "gr", 
    "exp": ["historical",], 
    "ensemble": "r6i1p1f1",
    "start_year": 2000,
    "end_year": 2022,
})
```

```py
forcing = GenericDistributedForcing.generate(
    dataset=cmip_dataset,
    start_time="2020-01-01T00:00:00Z",
    end_time="2020-01-01T00:00:00Z",
    shape="/home/bart/git/ewatercycle/src/ewatercycle/testing/data/Rhine/Rhine.shp",
)
```
</details>

sverhoeven committed:
Make mip required in dataset

sverhoeven committed:
More docs

sverhoeven committed:
Make expected recipe string consistent

sverhoeven committed:
Parse esmvaltool recipe output when non netcdf files are produced

+ more docs + fix tests

sverhoeven committed:
Got LisfloodForcing.generate(lisvap=None) working

Had to make Dataset.mip optional otherwise tdps var with mip=day throws esmvaltool error

sverhoeven committed:
Make MarrmotForcing.generate work

sverhoeven committed:
Fix tests for wflow and pcrglobwb

sverhoeven committed:
Add support for variable without temporal selection

sverhoeven committed:
give right dataset to lisvap

sverhoeven committed:
Correct expected orog var

sverhoeven committed:
Make mypy happy

sverhoeven committed:
Dont copy line number, >>> and ... from code blocks

sverhoeven committed:
More tests, flake8 fixes and docs

sverhoeven committed:
Tests for DefaultForcing + Remove discriminator from forcing classes

As you now have to specify which class you want to use to load a forcing.

sverhoeven committed:
Replaced with tests/src/base/test_forcing.py

sverhoeven committed:
Remove discriminator from forcing classes from tests

sverhoeven committed:
Make flake8 happier

sverhoeven committed:
Model validator should return otherwise object is set to None

sverhoeven committed:
Add test for MarrmotForcing.to_xarray()

sverhoeven committed:
Add generic forcing to user guide + custom forcing instructions +

prevent circular deps with generic forcings in entry points

sverhoeven committed:
More docs

sverhoeven committed:
Use Dataset object instead of dict in pre defined datasets

sverhoeven committed:
Tested GenericDistributedForcing with non-era dataset

sverhoeven committed:
Allow extra attributes in Dataset

'*' in forcing files was caused by version attribute in Dataset being silently forgotten.
By allowing extra attributes the generate file has version value instead of `*`.

sverhoeven commented:
> 

Your CMIP example now works and is part of docstring at https://ewatercycle--365.org.readthedocs.build/en/365/autoapi/ewatercycle/base/forcing/index.html#ewatercycle.base.forcing.GenericDistributedForcing

sverhoeven committed:
Tested CMIP6

sverhoeven committed:
Add dataset.version to expected recipes

sverhoeven committed:
Add changes to CHANGELOG + more docs + more todos

sverhoeven committed:
Fix tests with updated model names

sverhoeven committed:
Disable lisflood model tests as downloading parameter set too slow

sverhoeven commented:
See https://github.com/eWaterCycle/infra/blob/main/README.md#mount-dcache-on-local-machine how to mount ERA data locally.

BSchilperoort reviewed (approved):
I've successfully tested generating generic forcing for lumped & distributed data, both CMIP6 and ERA5 :+1: 

Thanks for adding the cmip support, having esmvaltool download the data on the fly makes testing the forcing generation a lot easier!

Peter9192 commented on src/ewatercycle/base/forcing.py:
```diff
@@ -123,107 +188,209 @@ def load(cls, directory: str | Path):
+                CMIP6: ~/climate_data/CMIP6
+            drs:
+                CMIP6: BADC
+                # If Synda is installed then use SYNDA instead of BADC
```
```suggestion                
```

With the updated download functionality, I don't think we want to recommend synda anymore

Peter9192 replied:
Might refer to detailed instructions here https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#user-configuration-file

And add somethink like: a default user config can be obtained by running `esmvaltool config get-config-user`

Peter9192 replied:
Maybe some of this documentation should go into the module docstring instead of the GenericDistributedForcing, as it is also relevant e.g. for the lumped forcing

sverhoeven replied:
Link and config getter added.

Moved to module docstring, with links from generic gorcing generate docstrings.

See 04fb15522ebe3bd7a63aaed6525f2505c160673d

Peter9192 commented on src/ewatercycle/base/forcing.py:
```diff
@@ -123,107 +188,209 @@ def load(cls, directory: str | Path):
+            rootpath:
+                CMIP6: ~/climate_data/CMIP6
+            drs:
+                CMIP6: BADC
```
```suggestion
                CMIP6: ESGF
```
Follow standard ESGF by default unless you're on a site-specific machine (as documented here: https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#user-configuration-file)

sverhoeven replied:
done

Peter9192 commented on src/ewatercycle/base/model.py:
```diff
@@ -83,6 +83,10 @@ def _check_parameter_set(self):
     def _make_bmi_instance(self) -> OptionalDestBmi:
         """Attach a BMI instance to self._bmi."""
 
+    # this has different signature than pymt see
+    # https://github.com/csdms/pymt/blob/9ef61a0010b4997d5a2b09e5d434371598291261/pymt/framework/bmi_setup.py#L77C21-L77C32
+    # where it is {}.items()
+    # TODO is this OK?
```
Might be better to conform to pymt, can you do that in the PR with the abstract model tests?

sverhoeven replied:
Done in 8c44672e6319bf60328cea08ee5c75400bb45aeb

Peter9192 commented on src/ewatercycle/esmvaltool/diagnostic/copier.py:
```diff
@@ -0,0 +1,55 @@
+        input_path = Path(input_file)
+        outfile = get_diagnostic_filename(input_path.stem, cfg, input_path.suffix[1:])
+        logger.info("Copying %s to %s", input_file, outfile)
+        # TODO copy or move?
```
For now copy is fine I think

sverhoeven replied:
copy it is

Peter9192 commented on src/ewatercycle/forcing.py:
```diff
@@ -42,10 +46,15 @@ def __repr__(self):
 _forcings: dict[str, Any] = {
-    entry_point.name: entry_point
-    for entry_point in entry_points(group="ewatercycle.forcings")  # /NOSONAR
+    "GenericDistributedForcing": GenericDistributedForcing,
+    "GenericLumpedForcing": GenericLumpedForcing,
```
I like this a lot; having defaults and adding plugins

Peter9192 commented on src/ewatercycle/plugins/hype/forcing.py:
```diff
@@ -124,3 +83,41 @@ def to_xarray(self) -> xr.Dataset:
         return ds
+
+
+def build_recipe(
```
Would it make sense to use model-specific names here? 

```suggestion
def build_hype_recipe(
```

sverhoeven replied:
suggestion applied.

Peter9192 commented on src/ewatercycle/plugins/hype/forcing.py:
```diff
@@ -124,3 +83,41 @@ def to_xarray(self) -> xr.Dataset:
+        .add_variable("tasmin", units="degC")
+        .add_variable("tasmax", units="degC")
+        .add_variable("pr", units="kg m-2 d-1")
+        .script("hydrology/hype.py")
```
I think it would be good to ship our own copies of the diagnostic scripts, or rather, have the plugins ship their own copies. Then we're not dependent on ESMValTool and everything that a model maintainer needs to maintain is in the same place.

sverhoeven replied:
can you create an issue for that?

Peter9192 replied:
Done #381

Peter9192 commented on tests/src/esmvaltool/test_models.py:
```diff
@@ -0,0 +1,65 @@
+from pathlib import Path
```
I found the name of this file `test_models` to be quite confusing. Can we make it more pydantic-specific?

sverhoeven replied:
I think you are confused these models with hydrological models 

I could rename the module to 
1. test_schema
2. test_recipe_models
3. test_io
4.  something else?

That models are implemented in pydantic should not be important to consumers so I dont like adding pydantic in the module name.

To make consistent we should also rename `ewatercycle.esmvaltool.models` .

I like schema.
@Peter9192 do you have a preference or suggestion?

Peter9192 replied:
I like schema too

sverhoeven replied:
schema it is

Peter9192 reviewed (approved):
Awesome work! I really love that we now have standard forcings for generic and lumped models, and that we can support much more ESMValTool functionality, most importantly any dataset. 

Since Bart already tried it out, I just quickly skimmed over the code and left a couple of small comments.

sverhoeven committed:
Implement PR suggestions

sverhoeven committed:
Rename build_recipe() to build_<modelname>_recipe()

sverhoeven committed:
Centralize esmvaltool config for download

sverhoeven committed:
Rename esmvaltool.models to esmvaltool.schema

When we talk about models we mean hydrological models.

sverhoeven commented:
Thanks for testing and reviewing.
